filesystem/iso9660.ksy

```diff
@@ -216,60 +94,867 @@ types:
+                            type: u4bi
+                          - id: dev_t_low
+                            type: u4bi
+                      rrip_px:
+                        doc-ref: rrip 4.1.1
+                        seq:
+                          - id: length
+                            type: u1
+                          - id: version
+                            type: u1
+                            valid: 1
+                          - id: file_mode
+                            type: u4bi
+                          - id: links
+                            type: u4bi
+                          - id: user
+                            type: u4bi
+                          - id: group
+                            type: u4bi
+                          - id: serial
+                            type: u4bi
+                            if: length >= 44
+                      rrip_re:
+                        doc-ref: rrip 4.1.5.3
+                        seq:
+                          - id: length
+                            type: u1
+                          - id: version
+                            type: u1
+                            valid: 1
+                      rrip_sf:
+                        doc-ref: rrip 4.1.7
+                        seq:
+                          - id: length
+                            type: u1
+                          - id: version
+                            type: u1
+                            valid: 1
+                          - id: virtual_file_size_high
+                            type: u4bi
+                          - id: virtual_file_size_low
+                            type: u4bi
+                          - id: table_depth
+                            type: u1
+                            enum: table_depth
+                        enums:
+                          table_depth:
+                            0x2: max_64kb
+                            0x4: max_64mb
+                            0x8: max_4gb
+                            0x10: max_1tb
+                            0x20: max_256tb
+                            0x40: max_64pb
+                            0x80: max_16eb
+                      susp_sl:
+                        doc-ref: susp 4.1.3.1
+                        seq:
+                          - id: length
+                            type: u1
+                          - id: version
+                            type: u1
+                            valid: 1
+                          - id: reserved
+                            type: b4
+                          - id: root
+                            type: b1
+                          - id: parent
+                            type: b1
+                          - id: current
+                            type: b1
+                          - id: continue
+                            type: b1
+                          - id: content
+                            type: str
+                            size: length - 5
```

Comment:
Move the 5 bytes into a subtype and replace `5` with `sizeof<the_subtype>`

Reply:
Well in the spec they give the total size of the block incl the header, so you have to reduce the size of the header.
And making breaking it into a subtype will just make more of a mess.

Reply:
> And making breaking it into a subtype will just make more of a mess.

Yes, I have to agree - it probably will. However, I have a better idea - you use these specific types in `header`:

https://github.com/kaitai-io/kaitai_struct_formats/blob/619e1ba426a52612905049c37adee9462e6f078c/filesystem/iso9660.ksy#L513-L535

and in fact, all of them start with `length` and `version`.

https://github.com/kaitai-io/kaitai_struct_formats/blob/619e1ba426a52612905049c37adee9462e6f078c/filesystem/iso9660.ksy#L958-L964

https://github.com/kaitai-io/kaitai_struct_formats/blob/619e1ba426a52612905049c37adee9462e6f078c/filesystem/iso9660.ksy#L558-L566

https://github.com/kaitai-io/kaitai_struct_formats/blob/619e1ba426a52612905049c37adee9462e6f078c/filesystem/iso9660.ksy#L645-L653

and so on. I checked that this pair appears in the beginning of all fields, with the only exception that `susp_unknown` doesn't have the `valid: 1` constraint on `version` (if this is intentional and not an oversight, we can easily get around it by using `if` or `valid/expr`).

It would be much better if you would extract these common fields right into `header` (because in fact, why not, it makes sense in my opinion - the names `length` and `version` sound like they totally fit into a header structure):

https://github.com/kaitai-io/kaitai_struct_formats/blob/619e1ba426a52612905049c37adee9462e6f078c/filesystem/iso9660.ksy#L508-L513

```diff
 header:
   seq:
     - id: signature
       type: u2be
       enum: signature
+    - id: length
+      type: u1
+    - id: version
+      type: u1
+      valid: 1
     - id: specific
```

The nice thing about it is that now you have the `length` available in the parent struct, so you can simply put the field `specific` in a substream. Then you will be able to throw out all the `size: length - 5` calculations and just use `size-eos: true`. Also, it won't allow the specific types to possibly get out of the designated substream and the parsing will fail on the first field that doesn't fit in there (not by ending up with a negative size from those `size: length - 5` calculations, which will also lead to an error, but fields with these calculations are not even present in all types - not sure if it's explicitly written in the reference spec that the `length` should be disregarded in some specific types, but if it's not, I'd definitely confine every specific type within that substream).

```diff
 header:
   seq:
@@ ... @@
     - id: specific
+      size: length - (signature._sizeof + length._sizeof + version._sizeof) # note: this `_sizeof` sum should probably be in a `value` instance
       type:
         switch-on: signature
         cases:
           'signature::aaip_attribute_list': susp_unknown # AL
@@ ... @@
           'signature::rrzf_zisofs': rrzf_zf # ZF
```

Now, you will just take each specific type, for example:

https://github.com/kaitai-io/kaitai_struct_formats/blob/619e1ba426a52612905049c37adee9462e6f078c/filesystem/iso9660.ksy#L722-L731

..., simply throw out `length` and `version`, and replace `size: length - 4` with `size-eos: true`:

```diff
 susp_pd:
   doc-ref: susp 5.2
   seq:
-    - id: length
-      type: u1
-    - id: version
-      type: u1
-      valid: 1
     - id: padding_area
-      size: length - 4
+      size-eos: true
```

You see how much better it is? You'll replace 18 identical `length`, `version` pairs in every single specific type with 1 common pair in `header`, you'll treat the `length` consistently and correctly, and eliminate the magic numbers in manual `size: length - 5` calculations. What more could we possibly wish for?

Reply:
"What more could we possibly wish for?"
For the feeling to go away that this PR is eternally unmergeable.

You can have it as-is and improve on it in your own time.